compiler/rustc_target/src/spec/mod.rs

```diff
@@ -1587,31 +1606,35 @@ impl Target {
             } );
             ($key_name:ident, list) => ( {
                 let name = (stringify!($key_name)).replace("_", "-");
-                if let Some(v) = obj.find(&name).and_then(Json::as_array) {
-                    base.$key_name = v.iter()
-                        .map(|a| a.as_string().unwrap().to_string())
-                        .collect();
+                if let Some(j) = obj.remove(&name){
+                    if let Some(v) = Json::as_array(&j) {
```

Comment:
I think this really wants to be removed (i.e. not warned against) if the value was definitely used. Right now keys that we expect to be arrays, but weren't used because they were not arrays are not warned against.

Reply:
Sounds good. I ended up splitting this into a separate warnings for unrecognized fields and recognized ones with the wrong json type.

Reply:
If you don't mind doing this, it would probably be better to adapt these to be a proper error instead – there's no reason for us to let this through, I feel. We already error for most of the other keys, even in cases where type is correct but the value isn't (e.g. `MergeFunctions`/`RelocModel`).

Reply:
Hm, actually I misread the code, we still proceed executing if the type of the value was wrong (but fail if it was right, but with a wrong value). This entire thing probably wants a refactor :sweat_smile:

Reply:
Yeah definitely agree that most of these should be hard errors instead. I made everything warnings in this case just to avoid breaking changes.

What would the process be for changing them to errors? Would it need to be part of the 2021 edition?

Reply:
Custom target definitions are an “unstable” feature, so introducing breaking changes here is okay without a process as extensive as an edition. Perhaps MCP at most, though I would say that even that is probably excessive in a situation where we're simply no longer accepting what was invalid in the first place.